packages/report-server/src/reportBuilder/output/output.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { Row } from '../types';
+import { outputBuilders } from './functions/outputBuilders';
+
+type OutputParams = {
+  type: keyof typeof outputBuilders;
+  config: unknown;
+};
+
+const output = (rows: Row[], params: OutputParams) => {
+  const { type, config } = params;
+
+  const outputBuilder = outputBuilders[type](config);
+  return outputBuilder(rows);
+};
+
+const buildParams = (params: unknown): OutputParams => {
+  const isOutPutBuildersType = (value: string): value is keyof typeof outputBuilders => {
+    switch (value) {
+      case 'default':
+      case 'matrix':
+        return true;
+      default:
+        return false;
+    }
+  };
```

Comment:
Yeah I'd say this is a bit too hacky, and a little brittle. It's always good to design the code in such a way that extensibility is easy. Ideally someone can just add a new ouput type to the `outputBuilders` object and then it can automatically be included in the validation logic, something like:
```suggestion
  const isOutPutBuilderType = (value: string): value is keyof typeof outputBuilders => {
    Object.keys(outputBuilders).includes(value);
  };
```

Reply:
Oh that is very smart 👍